src/btree/bt_curnext.c

```diff
@@ -892,21 +906,45 @@ __wt_btcur_next(WT_CURSOR_BTREE *cbt, bool truncating)
         /*
          * If we saw a lot of deleted records on this page, or we went all the way through a page
          * and only saw deleted records, try to evict the page when we release it. Otherwise
-         * repeatedly deleting from the beginning of a tree can have quadratic performance. Take
-         * care not to force eviction of pages that are genuinely empty, in new trees.
-         *
-         * A visible stop timestamp could have been treated as a tombstone and accounted in the
-         * deleted count. Such a page might not have any new updates and be clean, but could benefit
-         * from reconciliation getting rid of the obsolete content. Hence mark the page dirty to
-         * force it through reconciliation.
+         * repeatedly searching from the beginning of a tree can have quadratic performance. Take
+         * care not to force eviction of genuinely empty pages, in new trees.
          */
-        if (page != NULL &&
-          (cbt->page_deleted_count > WT_BTREE_DELETE_THRESHOLD ||
-            (newpage && cbt->page_deleted_count > 0))) {
-            WT_ERR(__wt_page_dirty_and_evict_soon(session, cbt->ref));
-            WT_STAT_CONN_INCR(session, cache_eviction_force_delete);
+        if (page != NULL) {
+            /*
+             * An obsolete visible stop timestamp could have been treated as a tombstone and
+             * accounted for in the deleted count. Such a page might not have any new updates and be
+             * clean, but could benefit from reconciliation getting rid of the obsolete content.
+             * Hence mark the page dirty to force it through reconciliation.
+             */
+            if (cbt->page_obsolete_deleted_count > WT_BTREE_DELETE_THRESHOLD ||
+              (newpage && cbt->page_obsolete_deleted_count > 0)) {
+                WT_ERR(__wt_page_dirty_and_evict_soon(session, cbt->ref));
+                WT_STAT_CONN_INCR(session, cache_eviction_force_obsolete_delete);
+            } else if (page->type != WT_PAGE_COL_FIX && total_skipped != 0 &&
+              F_ISSET(cbt, WT_CBT_ALL_DELETED_ITEMS) &&
+              session->txn->isolation != WT_ISO_READ_UNCOMMITTED) {
+                /*
+                 * A visible stop timestamp could have been treated as a tombstone and accounted for
+                 * page with no non-deleted updates. Evicting these pages could benefit search
+                 * performance. Generally, tombstones are ignored while searching the history store,
+                 * so we should not see any history store pages in this flow. Ignore the pages from
+                 * fixed length column store tables as they don't get any search performance benefit
+                 * and also sessions with READ UNCOMMITTED isolation, as their search can see the
+                 * results of an uncommitted transaction.
+                 *
+                 * Verify whether any deleted keys were skipped throughout the scan to find empty
```

Comment:
I don't understand this sentence. Empty page means no key at all on the page or keys are all deleted?

Reply:
Empty pages means no key at all on the page, updated the sentence for more clarity.